rust/lance-index/src/vector/ivf/shuffler.rs

```diff
@@ -302,28 +355,89 @@ impl IvfShuffler {
-            let lance_schema = reader
-                .schema()
-                .project(&[PART_ID_COLUMN])
-                .expect("part id should exist");
-
-            let mut stream = stream::iter(start..end)
-                .map(|i| reader.read_batch(i as i32, .., &lance_schema))
-                .buffer_unordered(16);
-
-            while let Some(batch) = stream.next().await {
-                let batch = batch?;
-                let part_ids: &UInt32Array = batch.column(0).as_primitive();
-                part_ids.values().iter().for_each(|part_id| {
-                    partition_sizes[*part_id as usize] += 1;
-                });
-            }
+
+            match self.output_version.as_str() {
+                "v1" => {
+                    let reader =
+                        FileReader::try_new_self_described(&object_store, &path, None).await?;
+                    let lance_schema = reader
+                        .schema()
+                        .project(&[PART_ID_COLUMN])
+                        .expect("part id should exist");
+
+                    let mut stream = stream::iter(start..end)
+                        .map(|i| reader.read_batch(i as i32, .., &lance_schema))
+                        .buffer_unordered(16);
+
+                    while let Some(batch) = stream.next().await {
+                        let batch = batch?;
+                        let part_ids: &UInt32Array = batch
+                            .column_by_name(PART_ID_COLUMN)
+                            .expect("Partition ID column not found")
+                            .as_primitive();
+                        part_ids.values().iter().for_each(|part_id| {
+                            partition_sizes[*part_id as usize] += 1;
+                        });
+                    }
+                }
+                "v2" => {
+                    let file = scheduler.open_file(&path).await?;
+                    let reader =
+                        Lancev2FileReader::try_open(file, None, Default::default()).await?;
+                    let mut stream = reader
+                        .read_stream(
+                            lance_io::ReadBatchParams::Range(start..end),
+                            1024,
+                            16,
+                            FilterExpression::no_filter(),
+                        )
+                        .unwrap();
+
+                    while let Some(batch) = stream.next().await {
+                        let batch = batch?;
+                        let part_ids: &UInt32Array = batch
+                            .column_by_name(PART_ID_COLUMN)
+                            .expect("Partition ID column not found")
+                            .as_primitive();
+                        part_ids.values().iter().for_each(|part_id| {
+                            partition_sizes[*part_id as usize] += 1;
+                        });
+                    }
```

Comment:
Seems like this could be potentially pulled out of the if statement if you changed it to:

```
let stream = if ...
```

Reply:
Both streams have different types...one is `BufferUnordered<Map<Iter<Range<usize>>, impl FnMut(usize) -> impl Future<Output = Result<RecordBatch, Error>>>>` and the other is `Pin<Box<dyn RecordBatchStream>>`

Reply:
Can you box the first one?  Just add `.boxed()` to it and it should also be `Pin<Box<dyn RecordBatchStream>>` (I think, might be wrong)

Reply:
I tried, but it's still different - `Pin<Box<dyn futures::Stream<Item = Result<arrow_array::RecordBatch, lance_core::Error>> + std::marker::Send>>`